admin/tools/provision/cloudmanager.py

```diff
@@ -0,0 +1,311 @@
+        instance = self.nova.servers.create(name=instance_name,
+                                            image=self.image,
+                                            flavor=self.flavor,
+                                            userdata=userdata,
+                                            key_name=self.key,
+                                            nics=self.nics)
+        # Poll at 5 second intervals, until the status is no longer 'BUILD'
+        status = instance.status
+        while status == 'BUILD':
+            time.sleep(5)
+            instance.get()
+            status = instance.status
+        logging.info("status: {}".format(status))
+        logging.info("Instance {} is active".format(instance_name))
+
+        return instance
+
+    def detect_end_cloud_config(self, instance):
+        """
+        Add clean wait for cloud-init completion
+        """
+        check_word = "---SYSTEM READY FOR SNAPSHOT---"
+        end_word = None
+        while not end_word:
+            time.sleep(10)
+            output = instance.get_console_output()
+            logging.debug("console output: {}".format(output))
+            logging.debug("instance: {}".format(instance))
+            end_word = re.search(check_word, output)
+
+    def nova_servers_delete(self, server):
+        """
+        Shut down and delete a server
+        """
+        server.delete()
+
+    def manage_ssh_key(self):
+        """
+        Upload ssh public key
+        """
+        logging.info('Manage ssh keys: {}'.format(self.key))
+        if self.nova.keypairs.findall(name=self.key):
+            logging.debug('Remove previous ssh keys')
+            self.nova.keypairs.delete(key=self.key)
+
+        with open(os.path.expanduser(self.key_filename + ".pub")) as fpubkey:
+            self.nova.keypairs.create(name=self.key, public_key=fpubkey.read())
+
+    def get_floating_ip(self):
+        """
+        Allocate floating ip address to project
+        """
+        i = 0
+        floating_ips = self.nova.floating_ips.list()
+        floating_ip = None
+        floating_ip_pool = self.nova.floating_ip_pools.list()[0].name
+
+        # Check for available public ip address in project
+        while i < len(floating_ips) and floating_ip is None:
+            if floating_ips[i].instance_id is None:
+                floating_ip = floating_ips[i]
+                logging.debug('Available floating ip found {}'.format(floating_ip))
+            i += 1
+
+        # Check for available public ip adress in ext-net pool
+        if floating_ip is None:
+            try:
+                logging.debug("Use floating ip pool: {}".format(floating_ip_pool))
+                floating_ip = self.nova.floating_ips.create(floating_ip_pool)
+            except novaclient.exceptions.Forbidden as exc:
+                logging.fatal("Unable to retrieve public IP: {0}".format(exc))
+                sys.exit(1)
+
+        return floating_ip
```

Comment:
I think there may be a race condition in this class. Here you return floating IP which is available at this moment, but by the time you assign it to a new instance this IP may become used by some other instance. The chance is probably small but it's non-zero.